bundles/auth/org.eclipse.smarthome.auth.jaas/src/main/java/org/eclipse/smarthome/auth/jaas/internal/JaasAuthenticationProvider.java

```diff
@@ -0,0 +1,130 @@
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+
+import org.eclipse.smarthome.core.auth.Authentication;
+import org.eclipse.smarthome.core.auth.AuthenticationException;
+import org.eclipse.smarthome.core.auth.AuthenticationProvider;
+import org.eclipse.smarthome.core.auth.Credentials;
+import org.eclipse.smarthome.core.auth.UsernamePasswordCredentials;
+
+/**
+ * Implementation of authentication provider which is backed by JAAS realm.
+ *
+ * Real authentication logic is embedded in login modules implemented by 3rd party, this code is just for bridging it to
+ * smarthome platform.
+ *
+ * @author Łukasz Dywicki - Initial contribution and API
+ * @author Kai Kreuzer - Removed ManagedService and used DS configuration instead
+ */
+public class JaasAuthenticationProvider implements AuthenticationProvider {
+
+    private String realmName;
+
+    @Override
+    public Authentication authenticate(final Credentials credentials) {
+        if (realmName == null) { // configuration is not yet ready or set
+            return null;
+        }
+
+        final String name = getName(credentials);
+        final char[] password = getPassword(credentials);
+
+        try {
+            LoginContext loginContext = new LoginContext(realmName, new CallbackHandler() {
+                @Override
+                public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+                    for (Callback callback : callbacks) {
+                        if (callback instanceof PasswordCallback) {
+                            ((PasswordCallback) callback).setPassword(password);
+                        } else if (callback instanceof NameCallback) {
+                            ((NameCallback) callback).setName(name);
+                        } else {
+                            throw new UnsupportedCallbackException(callback);
+                        }
+                    }
+                }
+            });
+            loginContext.login();
+
+            // TODO shall we call logout method on login context?
+            return getAuthentication(name, loginContext.getSubject());
+        } catch (LoginException e) {
+            throw new AuthenticationException("Could not obtain authentication over login context", e);
+        }
+    }
+
+    private Authentication getAuthentication(String name, Subject subject) {
+        return new Authentication(name, getRoles(subject.getPrincipals()));
+    }
+
+    private String[] getRoles(Set<Principal> principals) {
+        List<String> roles = new ArrayList<>();
+        for (Principal principal : principals) {
+            roles.add(principal.getName());
+        }
+
+        return roles.toArray(new String[roles.size()]);
```

Comment:
For current implementation your both points are valid, however there are different types of principals. Each JAAS login module might have it's own principal implementations which can mean group or role (ie. package `org.apache.karaf.jaas.boot.principal` contains `RolePrincipal`, `UserPrincipal`, `GroupPrincipal`). I planned to add in future some customization which would allow to filter out which principal instances should be used, then we can't rely on passed set instance size. However, this is longer run change which can be pushed after some security is introduced.